zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java

```diff
@@ -87,36 +89,51 @@ private String getData(String msg) {
     if (msg == null) {
       return null;
     }
-
-    Type[] types = Type.values();
-    for (Type t : types) {
-      String magic = "%" + t.name().toLowerCase();
-      if (msg.startsWith(magic + " ") || msg.startsWith(magic + "\n")) {
-        int magicLength = magic.length() + 1;
-        if (msg.length() > magicLength) {
-          return msg.substring(magicLength);
-        } else {
-          return "";
-        }
-      }
+    Type[] types = type.values();
+    TreeMap<Integer, Type> typesLastIndexInMsg = buildIndexMap(msg);
+    if (typesLastIndexInMsg.size() == 0) {
+      return msg;
+    } else {
+      Map.Entry<Integer, Type> lastType = typesLastIndexInMsg.lastEntry();
+      int magicLength = lastType.getValue().name().length() + 1; //add 1 for the %
+      int subStringPos = magicLength + lastType.getKey() + 1; // 1 for the last \w after magic
```

Comment:
Can someone look at this one and merge it if possible?
Eran

בתאריך שבת, 25 ביולי 2015, 22:51 מאת Felix Cheung <notifications@github.com
>:

> In
> zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterResult.java
> <https://github.com/apache/incubator-zeppelin/pull/164#discussion_r35486267>
> :
>
> > -      if (msg.startsWith(magic + " ") || msg.startsWith(magic + "\n")) {
> > -        int magicLength = magic.length() + 1;
> > -        if (msg.length() > magicLength) {
> > -          return msg.substring(magicLength);
> > -        } else {
> > -          return "";
> > -        }
> > -      }
> > +    Type[] types = type.values();
> > +    TreeMap<Integer, Type> typesLastIndexInMsg = buildIndexMap(msg);
> > +    if (typesLastIndexInMsg.size() == 0) {
> > +      return msg;
> > +    } else {
> > +      Map.Entry<Integer, Type> lastType = typesLastIndexInMsg.lastEntry();
> > +      int magicLength = lastType.getValue().name().length() + 1; //add 1 for the %
> > +      int subStringPos = magicLength + lastType.getKey() + 1; // 1 for the last \w after magic
>
> Cool!!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/incubator-zeppelin/pull/164/files#r35486267>.
>